File: algorithm/bucket-sort.js
/**
 * Bucket Sort Implementation in JavaScript
 * 
 * Bucket sort distributes values into buckets, sorts each bucket, then concatenates.
 * Works best when input is uniformly distributed.
 * 
 * Time Complexity: O(n + k) average, O(n²) worst case
 * Space Complexity: O(n + k)
 */

function bucketSort(arr, bucketCount = 10) {
    if (arr.length === 0) return arr;
    
    const min = Math.min(...arr);
    const max = Math.max(...arr);
    const bucketSize = (max - min) / bucketCount;
    
    // Create buckets
    const buckets = Array.from({ length: bucketCount }, () => []);
    
    // Distribute elements into buckets
    for (const num of arr) {
        const bucketIndex = Math.min(
            Math.floor((num - min) / bucketSize),
            bucketCount - 1
        );
        buckets[bucketIndex].push(num);
    }
    
    // Sort each bucket and concatenate
    const sorted = [];
    for (const bucket of buckets) {
        // Using insertion sort for each bucket (can use any sorting algorithm)
        insertionSort(bucket);
        sorted.push(...bucket);
    }
    
    return sorted;
}

function insertionSort(arr) {
    for (let i = 1; i < arr.length; i++) {
        const key = arr[i];
        let j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// Example usage
if (require.main === module) {
    const arr = [0.42, 0.32, 0.33, 0.52, 0.37, 0.47, 0.51];
    console.log('Bucket Sort:', bucketSort(arr));
}

module.exports = { bucketSort };

File: algorithm/bucket-sort.py
"""
Bucket Sort Implementation in Python

Bucket sort distributes values into buckets, sorts each bucket, then concatenates.
Works best when input is uniformly distributed.

Time Complexity: O(n + k) average, O(n²) worst case
Space Complexity: O(n + k)
"""


def bucket_sort(arr, bucket_count=10):
    if len(arr) == 0:
        return arr
    
    min_val = min(arr)
    max_val = max(arr)
    bucket_size = (max_val - min_val) / bucket_count
    
    # Create buckets
    buckets = [[] for _ in range(bucket_count)]
    
    # Distribute elements into buckets
    for num in arr:
        bucket_index = min(
            int((num - min_val) / bucket_size),
            bucket_count - 1
        )
        buckets[bucket_index].append(num)
    
    # Sort each bucket and concatenate
    sorted_arr = []
    for bucket in buckets:
        insertion_sort_bucket(bucket)
        sorted_arr.extend(bucket)
    
    return sorted_arr


def insertion_sort_bucket(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


if __name__ == "__main__":
    arr = [0.42, 0.32, 0.33, 0.52, 0.37, 0.47, 0.51]
    print("Bucket Sort:", bucket_sort(arr))

